### PR TITLE
feature(ntrip): send GGA position feedback to NTRIP caster for VRS support

### LINKED ISSUE #84: feature(ntrip): send GGA position feedback to NTRIP caster for VRS support
## User Story / Problem Statement

As a **developer using RTK GNSS on a rover**, I want the NTRIP client to send the rover's current GGA position back to the caster so that VRS (Virtual Reference Station) services provide geographically tailored RTCM3 corrections and achieve centimeter-level accuracy.

## Purpose / Background (Why)

`NTRIPClient` currently streams RTCM3 corrections from the caster to the GPS receiver via serial port. However, many NTRIP services — especially Japanese VRS networks (e.g. CORS-Japan / GNSS-NET) — require the rover to send its approximate position as an NMEA GGA sentence back to the caster over the same TCP connection, immediately after the HTTP handshake, and periodically thereafter (typically every ~10 s).

Without GGA feedback:
- VRS-type casters reject the connection outright, or
- The caster generates corrections for a default/wrong geographic area, reducing RTK fix quality or preventing RTK convergence entirely.

The `GNSSReader` already decodes RTK fix statuses (`RTK_FIXED` → `fix_quality=4`, `RTK_FLOAT` → `fix_quality=5`), so the infrastructure for consuming improved positions is in place. The missing piece is the feedback loop from rover → caster.

NTRIP v1 protocol: after the `ICY 200 OK` + headers, the TCP connection becomes bidirectional — the caster streams RTCM3 bytes while the client writes NMEA GGA sentences upstream on the same socket.

## Acceptance Criteria

- [ ] `NTRIPClient` accepts an optional `gga_provider: Callable[[], GGAData | None]` parameter.
- [ ] After the HTTP handshake succeeds, if a provider is given, the client formats and sends an initial GGA sentence over the TCP socket to the caster.
- [ ] The client re-sends a fresh GGA sentence every ~10 s while forwarding RTCM3 (interval configurable via `NTRIPConfig` or a constant).
- [ ] A new `format_gga(gga: GGAData) -> str` function in `sensing/nmea/` serializes `GGAData` to a valid NMEA `$GPGGA,...*XX\r\n` sentence (decimal degrees → DDMM.MMMM, XOR checksum).
- [ ] `gga_provider` is optional: when omitted (or returns `None`), behavior is identical to the current implementation (no regression).
- [ ] Unit tests cover `format_gga` (round-trip with `parse_gga`) and the GGA-sending logic in `NTRIPClient` (mock socket).
- [ ] All linting (`ruff`), type checking (`mypy --strict`), and tests pass.

## Out of Scope

- NTRIP v2 support (chunked HTTP, `/` sourcetable).
- Parsing or validating RTCM3 message content.
- Automatic reconnection on TCP drop.
- Sending GGA via the serial device back to the receiver (the receiver already gets corrections; this is about the TCP uplink to the caster).

## Success Metrics

- On a VRS-type NTRIP service, `GNSSData.gga.fix_quality` converges to `4` (RTK Fixed) within the normal RTK convergence window (~60 s) after enabling the client, where it previously stayed at `1` (standard GPS) or the caster rejected the connection.

## Dependencies

- `sensing/ntrip/client.py` — existing `NTRIPClient`
- `sensing/nmea/` — needs a new `format_gga` serializer (currently only a parser exists)
- `sensing/gnss/types.py` — `GGAData` is the input type for the provider callable

## Alternatives Considered

| Alternative | Rejected Because |
|-------------|------------------|
| Pass a pre-formatted GGA string instead of `GGAData` | Shifts formatting burden to callers; harder to test and validate checksum correctness centrally |
| Use a static hardcoded position from `NTRIPConfig` | Position becomes stale as the rover moves; VRS corrections degrade with distance from the declared position |
| Start a separate TCP connection just for GGA upload | Not how NTRIP v1 works; the protocol uses the single existing connection bidirectionally |

## Risks / Concerns

- Thread safety: `_forward` reads from the TCP socket while GGA writes go to the same socket. Python's `socket.sendall` is not thread-safe if called concurrently with `read1`. The GGA send must be coordinated with the read loop (e.g. a lock, or injected into the same loop thread).
- Some casters ignore GGA feedback entirely (non-VRS, single-base); the optional parameter ensures zero impact for those deployments.
- NMEA checksum must be correct; a bad sentence is silently discarded by the caster.

## Implementation Overview / Ideas (Optional)

1. **`sensing/nmea/formatter.py`** (new file, ≤ 100 lines):
   - `_decimal_to_nmea(degrees: float, is_lat: bool) -> tuple[str, str]` — converts decimal degrees to `DDMM.MMMM` + hemisphere indicator.
   - `_nmea_checksum(payload: str) -> str` — XOR of bytes between `$` and `*`.
   - `format_gga(gga: GGAData) -> str` — assembles and checksums a `$GPGGA` sentence from `GGAData`.

2. **`sensing/ntrip/client.py`** changes:
   - Add `gga_provider: Callable[[], GGAData | None] | None = None` to `NTRIPClient.__init__`.
   - Extract a `_gga_loop(sock, provider, cancel, interval)` helper that runs inside `stream()`, writing GGA to the socket on a timer alongside the `_forward` read loop. Use a `threading.Lock` on the socket write path.

3. **`sensing/ntrip/config.py`**: optionally add `gga_interval_seconds: float = 10.0` to `NTRIPConfig` for the resend interval.

## References

- NTRIP v1 Technical Paper: https://www.rtcm.org/differential-global-navigation-satellite-dgnss-standards
- NMEA 0183 GGA sentence format (NMEA-0183 standard §5.3.2)
- gpsd JSON protocol: https://gpsd.gitlab.io/gpsd/gpsd_json.html
- Existing implementation: `sensing/ntrip/client.py`, `sensing/nmea/gga.py`

### PR DESCRIPTION
**Pull request by IshitaTakeshi**
## Related Issue

Closes #84

## Context

Many NTRIP services -- especially Japanese VRS (Virtual Reference Station) networks -- require the rover to send its current position as an NMEA GGA sentence back to the caster over the same TCP connection. The caster uses this to generate RTCM3 corrections tailored to the rover's location. Without GGA feedback, VRS casters either reject the connection outright or produce corrections for the wrong geographic area, preventing RTK convergence.

`NTRIPClient` previously only forwarded RTCM3 from the caster to the serial device and never sent anything upstream. `GNSSReader` already supported RTK fix statuses (`fix_quality=4/5`), but that path was unreachable without VRS corrections flowing correctly.

## Changes

- **`sensing/nmea/formatter.py`** (new): `format_gga(gga: GGAData) -> str` serializes a `GGAData` to a valid `$GPGGA` sentence (DDMM.MMMM coordinate format, XOR checksum, `\r\n` terminated).
- **`sensing/nmea/__init__.py`**: exports `format_gga`.
- **`sensing/ntrip/config.py`**: adds `gga_interval_seconds: float = 10.0` to `NTRIPConfig`.
- **`sensing/ntrip/client.py`**:
  - New `_maybe_send_gga` helper -- sends a GGA sentence if the interval has elapsed.
  - New `_make_gga_sender` -- closure binding the socket and GGA provider callable.
  - `_forward` -- extended with `gga_sender` and `gga_interval` parameters.
  - `NTRIPClient.__init__` -- adds optional `gga_provider: Callable[[], GGAData | None] | None = None`.
  - No extra threads needed; TCP is full-duplex so reads and writes on the same socket are independent.
- **`tests/nmea/test_formatter.py`** (new): 9 tests covering `format_gga`, including a round-trip with `parse_gga` using Tokyo Tower coordinates.
- **`tests/ntrip/test_client.py`**: 5 new tests covering the GGA send path and all "no send" cases (absent provider, `None` return, missing coordinates).

## Type of Change

- [x] New feature (non-breaking -- `gga_provider` defaults to `None`, existing behaviour unchanged)

## Test Steps

1. `uv run --extra dev pytest tests/ -q` -- all 178 tests pass.
2. `uv run --extra dev ruff check sensing/nmea/formatter.py sensing/nmea/__init__.py sensing/ntrip/config.py sensing/ntrip/client.py` -- no errors.
3. `uv run --extra dev mypy sensing/nmea/formatter.py sensing/nmea/__init__.py sensing/ntrip/config.py sensing/ntrip/client.py` -- no issues.

## Checklist

- [x] Tests added for new behaviour
- [x] Linting passes
- [x] Type checking passes
- [x] Backwards compatible -- omitting `gga_provider` preserves existing behaviour exactly

## Review Focus

- `_maybe_send_gga` in `client.py`: verify the timer logic and that the `last_gga=0.0` initial value correctly triggers an immediate first send.
- `_degrees_to_nmea` in `formatter.py`: verify the DDMM.MMMM conversion is correct (especially the `%07.4f` format for sub-10-degree minutes).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sensing/ntrip/client.py [hunk 1]
```diff
@@ -81,6 +84,71 @@ def _write_all(serial: io.RawIOBase, data: bytes) -> None:
...
+    the full interval.
+
+    Args:
+        gga_sender: Callable that sends one GGA sentence to the caster socket.
+        cancel: Event that signals the loop to stop.
+        gga_interval: Seconds to wait between consecutive GGA sends.
+    """
+    gga_sender()
+    while not cancel.wait(gga_interval):
+        gga_sender()
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/ntrip/client.py]: This code no longer exists — `_gga_loop` and the `last_gga` pattern were replaced by `RepeatingTask` in a subsequent refactor (see `sensing/concurrency.py`). `RepeatingTask` calls the task once immediately on `__enter__`, then uses `Event.wait(interval)` to sleep between repeats, so the timing concern no longer applies. (Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778)

File: sensing/ntrip/client.py [hunk 2]
```diff
@@ -81,6 +84,71 @@ def _write_all(serial: io.RawIOBase, data: bytes) -> None:
...
+    Returns:
+        The started ``threading.Thread``, or ``None`` if no provider was given.
+    """
+    if gga_provider is None:
+        return None
+    gga_sender = _make_gga_sender(sock, gga_provider)
+    thread = threading.Thread(
+        target=_gga_loop, args=(gga_sender, cancel, gga_interval), daemon=True
+    )
+    thread.start()
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/ntrip/client.py]: Same as above — this code was removed in the refactor to `RepeatingTask`. No longer applicable. (Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778)

File: sensing/ntrip/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/ntrip/client.py]: The busy-loop concern is valid — addressed directly in `sensing/concurrency.py`: `RepeatingTask.__init__` now raises `ValueError` for `interval <= 0`. The `_gga_loop` code that prompted the comment was removed in the refactor to `RepeatingTask`. (Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778)

File: sensing/ntrip/config.py [hunk 3]
```diff
@@ -24,3 +24,4 @@ class NTRIPConfig:
     serial_device: str
     username: str = ""
     password: str = ""
+    gga_interval_seconds: float = 10.0
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/ntrip/config.py:29]: Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778 — `gga_interval_seconds` is now documented in the Args section of `NTRIPConfig`'s docstring.

File: sensing/ntrip/client.py [hunk 4]
```diff
@@ -136,12 +204,23 @@ class NTRIPClient:
 
     Args:
         config: NTRIP caster and serial device parameters.
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/ntrip/client.py:180]: Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778 — the `__init__` docstring now reads ```stream()``` consistently with the rest of the module.

File: sensing/concurrency.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    ) -> None:
+        """Cancel the background thread and block until it exits."""
+        self._cancel.set()
+        self._thread.join()
+
+    def _loop(self) -> None:
+        """Call the task immediately, then repeat until cancel is set."""
+        self._task()
+        while not self._cancel.wait(self._interval):
+            self._task()
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/concurrency.py:57]: Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778 — `RepeatingTask.__init__` now raises `ValueError` if `interval <= 0`, preventing the busy-loop entirely. The docstring's Raises section is updated accordingly.

File: sensing/nmea/formatter.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    """Return the two-digit uppercase hex XOR checksum of payload."""
+    value = reduce(lambda acc, c: acc ^ ord(c), payload, 0)
+    return f"{value:02X}"
+
+
+def _degrees_to_nmea(decimal: float, *, is_latitude: bool) -> tuple[str, str]:
+    """Convert decimal degrees to NMEA DDMM.MMMM / DDDMM.MMMM and hemisphere char."""
+    abs_deg = abs(decimal)
+    degrees = int(abs_deg)
+    minutes = (abs_deg - degrees) * 60.0
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply sensing/nmea/formatter.py]: Confirmed and fixed in 928368759c7cc6ae769b78903c8d0581f05cc778. `_degrees_to_nmea` now rounds minutes to 4 d.p. first, then normalises: if `minutes >= 60.0`, it subtracts 60 and increments `degrees`. This prevents the invalid `DD60.0000` output at boundary values.

File: tests/ntrip/test_client.py [hunk 7]
```diff
@@ -21,6 +22,24 @@
     "rtk.example.com", 2101, "test-mount", "/dev/ttyAMA5",
     username="user", password="pass",  # noqa: S106
 )
+# gga_interval_seconds=0 ensures GGA is sent on the first loop iteration.
+_CFG_GGA = NTRIPConfig(
+    "rtk.example.com", 2101, "test-mount", "/dev/ttyAMA5",
+    gga_interval_seconds=0.0,
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply tests/ntrip/test_client.py]: Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778 — changed to `gga_interval_seconds=10.0` and removed the stale comment. The tests remain valid because `RepeatingTask` always fires the task once immediately on `__enter__`, regardless of interval.

File: tests/concurrency/test_repeating_task.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        done = threading.Event()
+
+        def task() -> None:
+            calls.append(1)
+            if len(calls) >= 2:
+                done.set()
+
+        with RepeatingTask(task, interval=0.0):
+            assert done.wait(timeout=1.0)
+        assert len(calls) >= 2
```
[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply tests/concurrency/test_repeating_task.py]: Fixed in 928368759c7cc6ae769b78903c8d0581f05cc778 — changed to `interval=0.01` and tightened the assertion to `2 <= len(calls) < 100`. The upper bound guards against runaway iteration while still confirming the repeat behaviour works.
